src/v/cloud_storage/segment_meta_cstore.cc

```diff
@@ -1038,7 +1038,12 @@ void segment_meta_cstore::from_iobuf(iobuf in) {
     *_impl = serde::from_iobuf<segment_meta_cstore::impl>(std::move(in));
 }
 
-iobuf segment_meta_cstore::to_iobuf() {
-    return serde::to_iobuf(std::exchange(*_impl, {}));
+iobuf segment_meta_cstore::to_iobuf() const {
```

Comment:
This is re-encoding every segment, right?

IIRC one of the original ideas of the column store was that we could stream the in memory state directly for serialization (perhaps by flushing write buffers first?).  @Lazin do you have context on whether that should be safe?

Reply:
Basically yes. The problem with the current combination of serde + move only types is that serialization would consume the state. the easy solution is to create a copy as I did here.
I can work out how to remove re-encoding though, by lowering the operations and assembling the iobuf "manually", and adding fuzz testing to check for bugs

Reply:
ok so I got some time to experiment and the underlying issue is that iobuf.share non const, while base_manifest::serialize is const. so the simple solution of creating a cstore sharing the buffers just for the purpose of serialization can't be done. 
the other options are to create a deep copy of the column store without re-encoding, by copying the buffers, o to manually serialize the store based on some proxy structure like is done in partition_manifest (in the simple form it would still require some copying of iobuffers probably)